File: container-documentapi/src/vespa/documentapi/messagebus/messages/documentstate.cpp
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "documentstate.h"
#include <vespa/document/util/bytebuffer.h>
#include <vespa/documentapi/common.h>
#include <vespa/vespalib/objects/nbostream.h>
#include <vespa/vespalib/util/growablebytebuffer.h>

namespace documentapi {

DocumentState::DocumentState()
    : _timestamp(0), _removeEntry(false) {}

DocumentState::DocumentState(const DocumentState& o)
    : _gid(o._gid), _timestamp(o._timestamp), _removeEntry(o._removeEntry)
{
    if (o._docId.get() != 0) {
        _docId = std::make_unique<document::DocumentId>(*o._docId);
    }
}

DocumentState::DocumentState(const document::DocumentId& id, uint64_t timestamp, bool removeEntry)
    : _docId(new document::DocumentId(id)),
      _gid(_docId->getGlobalId()),
      _timestamp(timestamp),
      _removeEntry(removeEntry)
{
}

DocumentState::DocumentState(const document::GlobalId& gid, uint64_t timestamp, bool removeEntry)
    : _gid(gid), _timestamp(timestamp), _removeEntry(removeEntry) {}

DocumentState::DocumentState(document::ByteBuffer& buf)
    : _docId(), _gid(), _timestamp(0), _removeEntry(false)
{
    uint8_t hasDocId;
    buf.getByte(hasDocId);
    if (hasDocId) {
        vespalib::nbostream stream(buf.getBufferAtPos(), buf.getRemaining());
        _docId = std::make_unique<document::DocumentId>(stream);
        buf.incPos(stream.rp());
    }
    const char* gid = buf.getBufferAtPos();
    buf.incPos(document::GlobalId::LENGTH);
    _gid.set(gid);
    buf.getLongNetwork((int64_t&) _timestamp);
    uint8_t b;
    buf.getByte(b);
    _removeEntry = b > 0;
}

DocumentState&
DocumentState::operator=(const DocumentState& other)
{
    _docId.reset();
    if (other._docId) {
        _docId = std::make_unique<document::DocumentId>(*other._docId);
    }
    _gid = other._gid;
    _timestamp = other._timestamp;
    _removeEntry = other._removeEntry;
    return *this;
}

void DocumentState::serialize(vespalib::GrowableByteBuffer &buf) const
{
    if (_docId.get()) {
        buf.putByte(1);
        string str = _docId->toString();
        buf.putBytes(str.c_str(), str.length() + 1);
    } else {
        buf.putByte(0);
    }
    char* buffer = buf.allocate(document::GlobalId::LENGTH);
    memcpy(buffer, _gid.get(), document::GlobalId::LENGTH);

    buf.putLong(_timestamp);
    buf.putByte(_removeEntry ? 1 : 0);
}

} // documentapi

File: documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/DocumentState.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.documentapi.messagebus.protocol;

import com.yahoo.document.DocumentId;
import com.yahoo.document.GlobalId;
import com.yahoo.text.Utf8;
import com.yahoo.vespa.objects.Deserializer;
import com.yahoo.vespa.objects.Serializer;

/**
 * @author <a href="mailto:einarmr@yahoo-inc.com">Einar M R Rosenvinge</a>
*/
public class DocumentState implements Comparable<DocumentState> {
    private DocumentId docId;
    private GlobalId gid;
    private long timestamp;
    private boolean removeEntry;

    public DocumentState(DocumentId docId, long timestamp, boolean removeEntry) {
        this.docId = docId;
        this.gid = new GlobalId(docId.getGlobalId());
        this.timestamp = timestamp;
        this.removeEntry = removeEntry;
    }

    public DocumentState(GlobalId gid, long timestamp, boolean removeEntry) {
        this.gid = gid;
        this.timestamp = timestamp;
        this.removeEntry = removeEntry;
    }

    public DocumentState(Deserializer buf) {
        byte hasDocId = buf.getByte(null);
        if (hasDocId == (byte) 1) {
            docId = new DocumentId(buf);
        }
        gid = new GlobalId(buf);
        timestamp = buf.getLong(null);
        removeEntry = buf.getByte(null)>0;
    }

    public DocumentId getDocId() {
        return docId;
    }

    public GlobalId getGid() {
        return gid;
    }

    public long getTimestamp() {
        return timestamp;
    }

    public boolean isRemoveEntry() {
        return removeEntry;
    }

    public void serialize(Serializer buf) {
        if (docId != null) {
            buf.putByte(null, (byte) 1);
            docId.serialize(buf);
        } else {
            buf.putByte(null, (byte) 0);
        }
        gid.serialize(buf);
        buf.putLong(null, timestamp);
        buf.putByte(null, (byte)(removeEntry ? 1 : 0));
    }

    public int getSerializedSize() {
        int size = 0;
        if (docId != null) {
            size += Utf8.byteCount(docId.toString()) + 1;
        }
        size += GlobalId.LENGTH;
        size += 8;
        size += 1;
        return size;
    }

    @Override
    public int compareTo(DocumentState state) {
        int comp = gid.compareTo(state.gid);
        if (comp == 0) {
            if (docId != null) {
                if (state.docId != null) {
                    return docId.toString().compareTo(state.docId.toString());
                } else {
                    return 1;
                }
            } else if (state.docId != null){
                return -1;
            }
        }
        return comp;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof DocumentState)) return false;

        DocumentState that = (DocumentState) o;

        if (removeEntry != that.removeEntry) return false;
        if (timestamp != that.timestamp) return false;
        if (docId != null ? !docId.equals(that.docId) : that.docId != null) return false;
        return gid.equals(that.gid);
    }

    @Override
    public int hashCode() {
        int result;
        result = (docId != null ? docId.hashCode() : 0);
        result = 31 * result + gid.hashCode();
        result = 31 * result + (int) (timestamp ^ (timestamp >>> 32));
        result = 31 * result + (removeEntry ? 1 : 0);
        return result;
    }

    @Override
    public String toString() {
        return "DocumentState{" +
               "docId=" + docId +
               ", gid=" + gid +
               ", timestamp=" + timestamp +
               ", removeEntry=" + removeEntry +
               '}';
    }

}
